R/data_write.r

```diff
@@ -0,0 +1,242 @@
+#' @param data The data frame that should be written to a file.
+#' @param delimiter For CSV-files, specifies the delimiter. Defaults to `","`,
+#'   but in particular in European regions, `";"` might be a useful alternative,
+#'   especially when CSV-files should be imported into MS Excel.
+#' @param save_variable_labels Only applies to CSV files. If `TRUE`, variable
```

Comment:
This argument seems dangerous because data classes are different from the original after writing + reading data with `data_read()` and it can error if we import with other packages:
``` r
library(datawizard)

data(efc)
d <- data_filter(efc, filter = 1:5)
d$e42dep <- droplevels(d$e42dep)

tmp <- tempfile(fileext = ".sav")

# before writing
lapply(d, class)
#> $c12hour
#> [1] "numeric"
#> 
#> $e16sex
#> [1] "numeric"
#> 
#> $e42dep
#> [1] "factor"
#> 
#> $c172code
#> [1] "numeric"
#> 
#> $neg_c_7
#> [1] "numeric"

data_write(d, tmp, save_variable_labels = TRUE)
#> Preparing data file: converting variable types.

# after writing and reading
data_read(tmp, convert_factors = FALSE) |> lapply(X = _, class)
#> Reading data...
#> $c12hour
#> [1] "numeric"
#> 
#> $e16sex
#> [1] "numeric"
#> 
#> $e42dep
#> [1] "numeric"
#> 
#> $c172code
#> [1] "numeric"
#> 
#> $neg_c_7
#> [1] "numeric"

data.table::fread(tmp) |> lapply(X = _, class)
#> Warning in data.table::fread(tmp): Detected 1 column names but the data has 5
#> columns (i.e. invalid file). Added 4 extra default column names at the end.
#> Error in data.table::fread(tmp): embedded nul in string: [TRUNCATED]
```

Reply:
What would you suggest? Giving a warning message, or skip this argument entirely? Or, maybe instead of adding a row, we could add the variable label to the column name (despite leading to _very long_ names then).

Reply:
The problem is that if you export to csv with variable labels and give this file to someone else, this person will have to know that class detection will be messed up because of the variable labels. Even if they know this, they have to:

* read the file
* extract the row with the labels and save them elsewhere
* convert all columns back to their right class
* add extracted labels in the attributes of the data/var

which is a lot just to have variable labels. To me, if someone wants to export a csv while not losing the variable label entirely, it's better to export the labels in a separate csv/RDS file

Reply:
so I would remove this argument

Reply:
I changed it to `save_labels`, which creates an extra file with value and variable labels